rollup/rollup
# Fix UMD wrapper issues and refine wrappers

This PR contains:

Are tests included?
- [x] yes (*bugfixes and features will not be merged without tests*)

Breaking Changes?
- [ ] yes (*breaking changes will not be merged unless absolutely necessary*)

List any relevant issue numbers:

### Description

With the recent changes to the UMD wrapper, several bugs were introduced. This addresses this and includes a thorough test suite of 240 tests (of which 84 were broken, some event before the latest refactorings!) that actually runs the UMD wrapper with different combinations of options and different scenarios:

options:
* simple global variable or namespaced.variable.with@special/characters
* compact or non-compact mode
* iife noConflict method present or not

environments:
* node
* simulated amd
* simulated iife where this = global variable
* simulated iife strict mode where this = undefined but self = global variable
* simulated iife with pre-existing bundle variable (to test the noConflict method)

tests:
* no imports/exports
* named exports
* default exports
* import + named exports
* import + default export

All tests are green now and the wrapper could be further refined. Moreover, `this` will now have precedence over `self` in iife mode and the check has been moved into the wrapper code, e.g.

```js
(function (global, factory) {
	typeof exports === 'object' && typeof module !== 'undefined' ? factory(require('foo'), require('bar')) :
	typeof define === 'function' && define.amd ? define(['foo', 'bar'], factory) :
	(global = global || self, factory(global.foo));
}(this, function (foo) { 'use strict';

	console.log( foo.a );

}));
```

A similar treatment is done for the IIFE wrappers.
